### ISSUE TITLE
Revert statistics table back to 1.14 version

### ISSUE THREAD

**Issue by nemaara** [wesnoth/wesnoth#4793]
So I quite like to look at my stats in the stats table from time to time when playing campaigns, but in the 1.15 version, it's incredibly hard to read/understand the table as it is. This is what it currently looks like:

![image](https://user-images.githubusercontent.com/13598479/76154436-4d8cc780-60aa-11ea-9bf2-d48bae35a6ce.png)

In general, there's too many numbers floating around and I really don't like that I actually have to do arithmetic on the numbers in parentheses to get the actual value of damage I did or the number of hits I got. The colored numbers are actually the least important, and also the most convoluted, so I really don't understand why they're colored (to make them stand out the most) or why they're even there in the first place. It's just really hard to find the information I want to know (how much damage did I do versus how much damage was I supposed to do?), which is probably the most important stat for me (the player) to know.

I would really love it if we could just have the old 1.14 statistics table back, which for reference looks like this:

![image](https://user-images.githubusercontent.com/13598479/76154451-9e042500-60aa-11ea-847f-9797004eb471.png)

There's just less numbers and I think the formatting is just way easier to understand. I wouldn't mind keeping the hits stats from the 1.15 version, just presented in the same way as in the 1.14 version.

**Comment by ElderoZ** [comment]
I second that.

**Comment by Vultraz** [comment]
Definitely should do this before 1.16.

**Comment by Mawmoocn** [comment]
Hello!

The reason why we change damage is for consistency on how we show information.

Originally, only hits was supposed to be affected by the UI change.

We were looking for suggestions on how to interpret the data better, and that was the result.

I don’t mind if damage reverts back but please don’t include hits as they both measure differently.

(#4113 the tweaks for hit statistics)

Anyways, if you have suggestions and critisims for it, please tell us so we can improve on how we interpret this data better.

**Comment by nemaara** [comment]
I'm not doubting how you're interpreting the data nor how good your metrics are for analyzing it. This issue is about how clearly it's displayed for understanding exactly what the numbers say. In any case, I think the first thing I'm going to want to know as a player is how much damage I've actually dealt (and/or the number of hits I've made), even regardless of what the expected value is. The problem with the 1.15 display is that it doesn't show me that anywhere (I have to calculate it myself by adding up the numbers in parentheses, which also don't tell me what they are besides in a paragraphs long tooltip). Showing the percent first also feels odd to me, but that might be because there's 2 numbers in parentheses next to it, so all the extra symbols make it feel cluttered and hard to read. 

The colored number in the hits line is the least important one (because it's the one where you actually analyze the data), so even if we left it in, it should stand out less than all of the other values in the table (i.e. be not colored). Usually, though, actual stats analysis is left for outside apps or in some advanced stats menu, not for the base display. You could easily choose a lot of other ways to analyze how "lucky" the players got with hundreds of metrics, so I really don't see the value in putting this one in. Honestly as it is, I already don't feel that the hits count is that valuable to show, but it might be relevant if you're looking at it on a turn by turn basis, or for scenarios where you have lower hit counts. The extra data beyond that is just too much.

**Comment by Wedge009** [comment]
I vaguely recall these additions were done for a reason. Supposed to give more information with respect to players complaining about excessively bad luck (ie showing that it is or isn't actually to do with luck). Was it @sigurdfdragon or @jostephd who did this? Or maybe @mattsc as part of AI work?

**Comment by Wedge009** [comment]
I vaguely recall these additions were done for a reason. Supposed to give more information with respect to players complaining about excessively bad luck (ie showing that it is or isn't actually to do with luck). Was it @sigurdfdragon or @jostephd who did this? Or maybe @mattsc as part of AI work?

**Comment by soliton-** [comment]
It was done by @jostephd. Part of the discussion can be seen in #4113 as @Mawmoocn mentioned.

@nemaara IMO a regression is if a bug that was already fixed is re-introduced or perhaps some feature is unintentionally broken. AFAIK this works as intended.

I think an advanced stats menu for those interested in more details sounds like a good idea.

**Comment by Vultraz** [comment]
It's a regression in usability. We can argue whether more detailed statistics should be made available, but this presentation is not the way to go. Look at it. What information could even a statistician glean from it? It requires long, extensive explanatory tooltips to decipher.

**Comment by IskarJarak** [comment]
Have to second Vultraz (and nemaara and EoZ and whoever commented similarly on discord). I looked the window up on my recent play-throughs for some reason (I think I wanted to see how badly reloading affected my hit-statistics or something) and gave up confused because I could neither find what I wanted to know nor make sense of the new numbers (how new are the mouse-overs? Pretty sure they weren't in at that point).

For anyone not trained in statistics this is a big step-down from 1.14.

(Just for the record, I _am_ trained in statistics and would still need some time with those tool-tips to make sense of (all) the data...)

**Comment by jostephd** [comment]
> I really don't understand why they're colored

The important thing about the coloring is that the color scale is reversed between the two lines: in the first line, 20 would be red and 80 green; in the second line, 80 would be red and 20 green. The first line is red to green, the second line is green to red. This way, **green means "lucky" and red means "unlucky"** in both lines. 

The tooltips should document that. If they aren't clear, perhaps you could help to improve them.

> I think the first thing I'm going to want to know as a player is how much damage I've actually dealt (and/or the number of hits I've made), even regardless of what the expected value is.

You're welcome to propose specific changes to the dialog along these lines.

Personally, I propose to remove the parentheses and their contents entirely. I realize this will remove the total/expected info which you said you want, but it will make the dialog less "scary" to novice users, which is your greater concern, if I understand correctly. Removing the parentheses and their content would leave in the dialog just percentages, or percentages and p-values, which is easier to explain. We could still have the expected and actual figures available somewhere, of course.

The thing is, I honestly think that if we have to choose between the percentage and the total (which one of them to place in the spotlight), we should choose the former. Compare saying "My HttT playthrough inflicted 24000 HP" to saying "My HttT playthrough had -10% inflicted". The second is a lot more informative. That's precisely why the dialog emphasizes it. In general, the dialog is designed to emphasize statistics (such as the p-value and the distance from the mean) that can be used to assess whether the result of the scenario was due to luck or skill.

> The colored number in the hits line is the least important one ... Honestly as it is, I already don't feel that the hits count is that valuable to show, but it might be relevant if you're looking at it on a turn by turn basis, or for scenarios where you have lower hit counts. The extra data beyond that is just too much.

Here we disagree. I think the colored part (the p-value) is actually the most interesting part of the dialog, because it's the only one that tells you whether luck was on your side or not.

**Comment by jostephd** [comment]
> It's a regression in usability. We can argue whether more detailed statistics should be made available, but this presentation is not the way to go. Look at it. What information could even a statistician glean from it? It requires long, extensive explanatory tooltips to decipher.

To a statistician, the explanation is a single sentence: "The colored part is the p-value of the observed result". In English, it's the probability of inflicting at least as many hits as had in fact been inflicted (if the result is 10 hits actual out of 8 expected, then it's the a priori probability of inflicting 10 or more hits).

The tooltip attempts to explain that in simpler terms for people without math background. Even a five years old kid should be able to understand the "green = lucky, red = unlucky" part of things.

**Comment by jostephd** [comment]
> Have to second Vultraz (and nemaara and EoZ and whoever commented similarly on discord). I looked the window up on my recent play-throughs for some reason (I think I wanted to see how badly reloading affected my hit-statistics or something) and gave up confused because I could neither find what I wanted to know nor make sense of the new numbers (how new are the mouse-overs? Pretty sure they weren't in at that point).

I think they were there since before the branch was merged to master but were improved after the merge. In any case, as has been said before: if anyone has specific changes to propose (either to the UI or to the documentation), please propose them. Otherwise we'll be talking in circles.

**Comment by shikadiqueen** [comment]
Just wanted to give my two cents and say that the UX for this dialog is embarrassingly poor both *before* and *after* the changes. It feels like it's intended to be some sort of insiders-only knowledge display.

**Comment by nemaara** [comment]
I agree with Iris. I don't think the 1.14 version is that great either, but the 1.15 version is even more convoluted and feels like something I (as the player) am not really expected to look at/deal with.

I think the best way to deal with this might be to only display the actual/expected values and leave off everything else, including the percentages. Any data interpretation can be left for advanced stats or for people interested in it to just calculate it themselves. I'm pretty sure all we need to show is the raw data (how much damage did the player do vs. what was the expected damage).

**Comment by jostephd** [comment]
> Any data interpretation can be left for advanced stats or for people interested in it to just calculate it themselves.

It's not possible to compute the p-values from the displayed information; they depend on the exact histogram of hits and misses, not just on the totals.

Anyway, I think we're going in circles. I think the p-values are important/useful and shouldn't be dropped and @nemaara insists that they are unimportant/confusing and should be the first thing removed. Anyone has a compromise proposal to suggest? Anyone at all?

**Comment by Ja-MiT** [comment]
It looks to me like the problem is that the UI was designed by people focused on statistics, rather than people focused on user experience. Both versions look like they were designed as homework to impress the teacher, rather than as help to enlighten the other students. Technically good, but rather dry. For a better user experience, focus on the 90% case first, which is a player with no specialized statistics background. Make the information readily accessible in this case, with extra information relegated to less prominent locations. (If the average player sees nothing recognizable, chances are good that the window will be closed with no tooltips seen.)

My view is that for the 90% case, the percentages are the most important numbers. Knowing that I did 12% more damage than "average" is easily comprehended and gives a quick snapshot of how I'm doing. Second to this, I would put a measure of lucky vs. unlucky. From a statistical perspective, this measure of luck is more important than the percentage, but it's harder to grasp without understanding statistics. One way to "soften the blow" would be to express luck as a color, as done in 1.15. (A numerical value should be available somewhere, though, as one should not rely on color alone to convey key information.)

My lowest priority would be "show your work", meaning showing the statistics upon which the above numbers are based. This is part trust-building and part letting players crunch the numbers themselves (if so inclined). (It seems to me that the 1.14 and 1.15 dialogs made showing the work the top priority instead of the bottom one.) *Keep in mind that the work does not have to all be shown directly in the window; we have tooltips that can convey additional, more in-depth information.*

---

I would put the percentage first in each stat block. (Ideally, this would be a slightly larger font than the rest of the text in the window.) Apply the "luck color" to this number. It's the first thing people see, so give it the two important bits of information. I do admit that there is a potential for confusion here: the color and the number are not directly related. A "-49%" might be neutrally colored early in a scenario but be rather green later after a large number of combats had been processed. Still, I think it's worth a try to see how it's received.

After the percentage, I'd put *some* supporting information in parentheses (ideally in a smaller font as well). For supporting information, I would include one piece of information for the 90% and one piece of information from the more hardcode stats, perhaps the ratio `actual /  expected` (as in 1.14) and the variance. While the p-value is useful, the variance is easier to explain and often gives similar information. In addition, the p-value is a percentage, so having it near the other percentage could confuse the 90%. More details could be available via tooltips. As an example:

    -19% (37 / 45.5 v 12.2)

I would give the percentage a tooltip with three parts. 1) State that the percentage measures how far the actual value is from the expected. 2) Explain that the color measures luck without getting too technical. 3) Mention the p-value for this particular stat (so that color is not the only way to convey this information).

A separate tooltip for the parenthetical part could be a place to list whatever statistics seem relevant but would be too "scary" for the main display. Actually, I might do one tooltip for the ratio, and a different tooltip for the variance. My plan might be for the former to give more of a layman's explanation while the latter could be geared towards statisticians.

Hmm... are different font sizes feasible? (I forget what the Wesnoth GUI supports easily.) If not, maybe show just the colored percentages in the window and relegate everything else to tooltips or some other initially-hidden state.

---

Sorry, my idea is petering out here. No time to do a mock-up or to look into the code. I guess my basic proposal for compromise is to initially hide the more technical stuff instead of getting rid of it completely.

**Comment by shikadiqueen** [comment]
> Keep in mind that the work does not have to all be shown directly in the window; we have tooltips that can convey additional, more in-depth information.

Or alternatively it could be hidden behind a toggle button intended for the more mathematically-oriented players.

> Hmm... are different font sizes feasible?

Yes.

**Comment by IskarJarak** [comment]
Ja-MIT's suggestion goes thematically in the right direction, but I have _no_ idea what you want there with the variance. That is as detrimental as the p-value for "normal" users. At a maximum, having % (actual / expected) would maybe work, everything more is incomprehensible number salad that, like you said, turns people away before they even discover the hover option (and that not only goes for the kind of information displayed, but even more so **for the amount of numbers listed** one after the other without any visible form of an identifier!).
I _do_ support having the more complex/nuanced numbers around but, as was suggested, there are a number of ways to have them not in the users face - a button for toggle/an additional window, or at least hide them in the hover (I kind of like how it breaks down hits by hit-chance as **supplemental** information currently).

If I remember I will ask my media informatics friends later today for suggestions, they are better than me with UI design.

**Comment by soliton-** [comment]
Perhaps it would be better to seperate probability stats from the simple stats that have nothing to do with probability. Probabilities are just inherintly difficult to understand and should just be shown in a more detailed dialog or via extending the initial dialog via some button or whatever.

The thing @nemaara mentioned that they want to know first/most when opening the statistics dialog is
> how much damage did I do versus how much damage was I supposed to do?

I think we should focus on the first trivial part here and just show damage done/taken as simple raw absolute values. No hint as to whether that is near the expected value or how lucky it was. Perhaps some more simple stats would be fun, like how many attacks were done/received...

Then for the second part "how much damage was I supposed to do?" which is a much more difficult question with no easy answers there would be an extended dialog that tries to explain the more advanced statistical analysis for those interested. For those not interested the initial dialog could just feature a mysteriously colored number or whatever trivial display we can come up with that shows our best guess at how lucky the player was.

**Comment by jostephd** [comment]
Another proposal: How about showing _either_ "Overall" _or_ "This turn", but not both at the same time?

**Comment by jostephd** [comment]
Using the statistics in @nemaara's screenshot, that could look like this:

| Hits (overall) | Inflicted | Taken |
| -------------- | --------- | ----- |
| Expected       | 4.9       | 3.7   |
| Actual         | 4         | 2     |
| Difference (%) | -18%      | -46%  |
| Luck score     | 32.5      | 12.6  |

Or we could do combine the damage and hits table, like this:

| | Damage (overall) |  | Hits (overall) |  |
|-|-|-|-|-|
|                | **Inflicted**        | **Taken** | **Inflicted**      | **Taken** |
| Expected       | 45.5             | 17.7  | 4.9            | 3.7   |
| Actual         | 37               | 9     | 4              | 2     |
| Difference (%) | -19%             | -49%  | -18%           | -46%  |
| Luck score     | —                | —     | 32.5           | 12.6  |

This mock-up looks like a salad, I know, but in-game we can make the percentages bold and keep the luck score colored, and then it might be more user friendly?

**Comment by jostephd** [comment]
Also, the "luck score" could be simplified by calculating it as probability of <= in inflicted but as probability of >= in taken (this is just the calculation under the hood, not something we'd say in the UI!), so both scores behave as "larger numbers is better".

**Comment by Mawmoocn** [comment]
Originally, I proposed hits as an advance feature because I cannot simplify it to the point of being understood without the help of tooltips.

We tried to simplify it to be understood by new players with the help of tooltips, that RNG is not very unfair but within the expected margin (unless you save-load), so they can probably concentrate on the game and not on the statistics itself.


Minus/Plus (-/+) separator was used to highlight the difference between large numbers.

For example:

`3795 / 3995.9 -5%`

The difference would be `200` if the result is lower than the expected amount, it is converted to:

`-5% (3995.9 - 200)`

Though that’s to simplify numbers but in reality what I wanted to do was:

`-5% (3795 - 3995.9 = 200)`
or
`3795 / 3995.9 (-200) -5%`
or
`3795 - 3995.9 (200) -5%`

and it was too long so I suggested the shoter option because I wanted to reduce the clutter of numbers and simplify the process.

The reason why I used Minus/Plus (-/+) is so that you would know how great or bad your rolls are for the current scenario, because I wanted to show the outright truth if it had gone bad or worse and the difference will always be within the margin of error or what we call CTH so that you can _probably_ focus on doing _better_.

The other reason why I re-positioned the percent was to reduce confusion when both percentages are aligned with each other and include the “luck” percentage.

It’s not surprising if it _can be_ confusing, because I wanted more input on how we can show it.... 

To be honest, I’m really looking for a better solution than what I had suggest.

I tried to simplify it to be understood by new players, that RNG is not very unfair but within the expected margin (unless you save-load), so they can concentrate on the game and not on the statistics itself.

**Comment by Mawmoocn** [comment]
> For a better user experience, focus on the 90% case first, which is a player with no specialized statistics background.

That was the plan, that’s why we opened up #4113 to tweak it for better UI experience.


> Make the information readily accessible in this case, with extra information relegated to less prominent locations.

We tried it on tool tips to explain how it works.

This can be improved better, we don’t know how so we’re stuck on tooltips at the moment because it’s more simple.

> My view is that for the 90% case, the percentages are the most important numbers. Knowing that I did 12% more damage than "average" is easily comprehended and gives a quick snapshot of how I'm doing.


This is true for short games. Long games will eventually even out the difference due to calculations that are adjusted to CTH.

So for long games, -5%, 0%, +5% will probably be common to see if you use your overall campaign statistics. (Except for special exceptions.)


> I would put a measure of lucky vs. unlucky.

This is very hard if not impossible, because probabilities are not fixed, as stated before, long games will even out any lucky circumstances you’d encounter.

Unless we can find a way to highlight luck without disregarding other forms of “luck”, it’ll be hard to contextualize luck without being seen as “unfair” (consistent).

RNG (or probability) by default, would never be consistent until a result has been made.



In reality this only a fragment of statistics that I wanted to see and hopefully we can highlight lucky hits and damage in the upcoming years due to it’s perceived complexity.

**Comment by Elvish-Hunter** [comment]
Just my two cents.
When I saw the mock-ups, the first thing I though was *"What does 'luck score' even mean?"*; then I noticed it refers to the *p-value*, but as it is it looks more like an unexplained magic number.
So, IMO it should either be removed (maybe this is the best option, since most people don't know enough about statistics to make any use of it) or called p-value (so anyone who wants to search for it and learn something can do so), but it shouldn't just stay as a mysterious "luck score".

**Comment by jostephd** [comment]
> So for long games, -5%, 0%, +5% will probably be common to see if you use your overall campaign statistics. (Except for special exceptions.)

Yes, that's how it is for the percentages. For example, if you open [the HttT final scenario replays thread](http://r.wesnoth.org/t10271) and look in the "All scenarios" statistics of the replays, you'll see most people have an overall actual/expected ratios in the range you mention. However, the p-values don't cluster this way. (By definition, about 40% of replays will have a p-value of 40 or less.) That's one of the things that make the p-value interesting to players.

**Comment by jostephd** [comment]
@Elvish-Hunter Thanks for the feedback. Happy to change the label to whatever people like, and we can put explanations in the tooltip (and/or in the in-game help). For example, we could call that row `Chance of a worse outcome` with a tooltip, maybe something like this:

> The probability of inflicting this many hits (4) or fewer.
>
> This is a measure of how lucky this side was. For example, if you finish a
> scenario with less than 5% inflicted and more than 95% taken, then you were
> very unlucky.
>
> The value is colored accordingly: red shades mean you were unlucky and green
> shades mean you were lucky.

and for the other cell in that row of the table:

> The probability of taking this many hits (2) or fewer.
>
> This is a measure of how lucky this side was. For example, if you finish a
> scenario with less than 5% inflicted and more than 95% taken, then you were
> very unlucky.
>
> The value is colored accordingly: red shades mean you were unlucky and green
> shades mean you were lucky.

@nemaara could probably improve these tooltips by miles, if everyone thinks the [split table mockup](https://github.com/wesnoth/wesnoth/issues/4793#issuecomment-597734170) is acceptable.

I couldn't work the terms "a priori" and "p-value" into the description within the confines of a tooltip. These are mathematical terms, so I guess they should only be mentioned in a help/wiki page that's specifically about the mathematical/randomness aspect of wesnoth, rather than in the main UI, since not all players want math terms in their front-line documentation :grin:

**Comment by Ja-MiT** [comment]
> > For a better user experience, focus on the 90% case first, which is a player with no specialized statistics background.
> 
> That was the plan, that’s why we opened up #4113 to tweak it for better UI experience.

Sorry, @Mawmoocn , but I don't see evidence of that plan there. Attempts to improve the UI, yes, but focusing on the 90%, no. The attempt may well have been there, and I believe the intentions were good,  but in my view the end result lost that focus.

> > My view is that for the 90% case, the percentages are the most important numbers. Knowing that I did 12% more damage than "average" is easily comprehended and gives a quick snapshot of how I'm doing.
> 
> This is true for short games. Long games will eventually even out the difference due to calculations that are adjusted to CTH.
> 
> So for long games, -5%, 0%, +5% will probably be common to see if you use your overall campaign statistics. (Except for special exceptions.)

You say this as if it is a negative. Why is that? What's wrong with seeing the law of large numbers in action?

**Comment by Mawmoocn** [comment]
@Ja-MiT

> Attempts to improve the UI, yes, but focusing on the 90%, no.

Okay I agree it isn’t “90%” but I hoped to simplify the learning process and not provide a golden spoon to everyone (including myself).

The result was what you see today because we thought it could be improved _later on_.

I know this is a problem and we hope to solve things together.

We need help.

> You say this as if it is a negative. Why is that? What's wrong with seeing the law of large numbers in action?

Sorry, I didn’t mean that way, I wanted to say that having another metric to measure things is a way to supplement the lack of detailed analysis and provide a simplified/summarized info to cut things short.

[This was my full proposal](r.wesnoth.org/t49785). (It's not polished but discussions and concessions are made along the way.)

To cut the long story short, I wanted a clear picture to know if CTH was within the expected range and where the mass of battle took place by using CTH.

(The (hits) statistics you see at the present time, we agreed on a simplified version derived from the (proposed) complex data. Complex data composes of different CTH values to show them separately without adding them all up. It only lists the CTH values that you and your foe had made in your play.)

I like seeing large numbers and I wanted to know the context behind it.

We're open to for more negative and positive feedback. I'm very thankful for knowing these issues so we can subjectively tackle the problem in a prudent manner.

**Comment by nemaara** [comment]
Yeah so based on all the discussion so far and talking about it with some people, I guess the expected value isn't something we should expect players to know about. So I'm thinking we should only give the percentage difference (how much better the player did than expected, without explaining that we calculate it vs. the expected value), and the actual raw damage they did (which I still think is relevant). With 2 numbers, it'll look a lot cleaner than it does now, and I thiiiink those won't be too complicated for anyone to really be confused about.

**Comment by Wedge009** [comment]
As a non-statistician (though I was quite proficient at mathematics in my younger years) and as a player I'm more inclined to go with the simpler 1.14 presentation as well (I haven't properly played anything on 1.15.x yet so I did not even know this was a new feature). However issues like #1378 (and maybe that's where some of these changes in 1.15 have come from) suggests that there are at least some people who want that extra information.

I haven't gone through the entire essay that is this thread, but I wonder if there's a possibility of having the simpler 1.14 statistics as the 'standard' version and the 1.15 statistics available as an 'advanced' version or some such. (Maybe someone else has already suggested this, I don't know.)

**Comment by soliton-** [comment]
Given #4113 was completed is there anything left to do here?
